modules/gapi/test/gapi_opaque_tests.cpp

```diff
@@ -0,0 +1,257 @@
+    auto out = ThisTest::FillMat::on(in, depth, chan, sz);
+
+    cv::GComputation c(cv::GIn(in), cv::GOut(out));
+    c.apply(cv::gin(fill), cv::gout(mat), cv::compile_args(cv::gapi::kernels<OCVFillMat>()));
+
+    auto diff = cv::Mat(sz, CV_MAKETYPE(depth, chan), cv::Scalar(fill)) - mat;
+    EXPECT_EQ(cv::countNonZero(diff), 0);
+}
+
+TEST(GOpaque, TestOpaqueBetween)
+{
+    cv::Size sz = {50, 50};
+    int depth = CV_8U;
+    int chan = 1;
+    cv::Mat mat_in = cv::Mat::zeros(sz, CV_MAKETYPE(depth, chan));
+    cv::Mat mat_out = cv::Mat::zeros(sz, CV_MAKETYPE(depth, chan));
+
+    cv::GMat in, out;
+    auto betw = ThisTest::GeneratePoint::on(in);
+    out = ThisTest::PaintPoint::on(betw, depth, chan, sz);
+
+    cv::GComputation c(cv::GIn(in), cv::GOut(out));
+    c.apply(cv::gin(mat_in), cv::gout(mat_out), cv::compile_args(cv::gapi::kernels<OCVGeneratePoint, OCVPaintPoint>()));
+
+    int painted = mat_out.at<uint8_t>(42, 42);
+    EXPECT_EQ(painted, 77);
+}
+
+TEST(GOpaque, TestOpaqueCustomOut2)
+{
+    cv::Mat input1 = cv::Mat(52, 52, CV_8U);
+    cv::Mat input2 = cv::Mat(42, 42, CV_8U);
+    std::string str = "opaque";
+    std::string str2 = str;
+    std::reverse(str2.begin(), str2.end());
+
+
+    ThisTest::MyCustomType out1, out2;
+
+    cv::GMat in1, in2;
+    auto out = ThisTest::GenerateOpaque::on(in1, in2, str);
+
+    cv::GComputation c(cv::GIn(in1, in2), cv::GOut(std::get<0>(out), std::get<1>(out)));
+    c.apply(cv::gin(input1, input2), cv::gout(out1, out2), cv::compile_args(cv::gapi::kernels<OCVGenerateOpaque>()));
+
+    EXPECT_EQ(out1.num, input1.size().width * input1.size().height);
+    EXPECT_EQ(out1.s, str);
+
+    EXPECT_EQ(out2.num, input2.size().width * input2.size().height);
+    EXPECT_EQ(out2.s, str2);
+}
+
+TEST(GOpaque_OpaqueRef, TestMov)
+{
+    // Warning: this test is testing some not-very-public APIs
+    // Test how OpaqueRef's mov() (aka poor man's move()) is working.
+
+    // a helper class to verify move semantics in GOpaque
+    struct Foo
+    {
+        char* cstring;
+
+        Foo(const char* s = "")
+        : cstring(nullptr)
+        {
+            if (s) {
+                std::size_t n = std::strlen(s) + 1;
+                cstring = new char[n];
+                std::memcpy(cstring, s, n);
+            }
+        }
+
+        ~Foo()
+        {
+            delete[] cstring;
+        }
+
+        Foo(const Foo& other)
+        : Foo(other.cstring)
+        {}
+
+        Foo(Foo&& other) noexcept
+        : cstring(exchange(other.cstring, nullptr))
+        {}
+
+        Foo& operator=(const Foo& other)
+        {
+             return *this = Foo(other);
+        }
+
+        Foo& operator=(Foo&& other) noexcept
+        {
+            std::swap(cstring, other.cstring);
+            return *this;
+        }
+    };
+
+    // compare 2 instances of Foo
+    auto foo_are_eq = [](const Foo& a, const Foo& b, std::size_t n){
+        return strncmp(a.cstring, b.cstring, n);
+    };
```

Comment:
Why craft this when we could use a regular std::string to check move semantics?

Reply:
True, it was over-complicated, fixed, thanks